### ISSUE TITLE
REST Client causes out of memory even if close is called.

### ISSUE THREAD

**Issue by pzima** [eclipse-ee4j/jersey#5486]
When using REST Client and requesting response in XML there is a memory leak. Closing of the client doesn't help. For JSON the problem is not present. See the attached test case that leads to out of memory
[JerseyMemoryTest.txt](https://github.com/eclipse-ee4j/jersey/files/13546405/JerseyMemoryTest.txt)
. Used jersey-client version was 3.1.2, but it fails with 3.1.4 too.

**Comment by plesingr-aeb** [comment]
This causes us big problems in the production - our servers must be restarted frequently. Could somebody suggest us at least a workaround (manual cleanup)? I tried to analyze the heap dump but unfortunately i was not able to detect the original cause yet. It seems that `ClientRuntime` instances hang somewhere in
`org.glassfish.hk2.internal.PerThreadContext$PerContextThreadWrapper.table(SystemDescriptor -> SecureSaxParserFactory)`.

**Comment by senivam** [comment]
As per the description, you would like to use one client for many requests. But in the reproducer you are creating as much clients as requests.

You can perfectly create one client for all requests like:

```
    @Test
    public void testMemoryConsumption() {
        int max = 100_000_000;
        try (final Client client = ClientBuilder.newClient()) {
            for (int i = 0; i < max; i++) {
                client.target(getBaseUri())
                        .path("memTest")
                        //.request(MediaType.APPLICATION_JSON)
                        .request(MediaType.APPLICATION_XML_TYPE)
                        .get(MemTestDto.class);
                if (i % 1_000 == 0) {
                    System.out.println("attempt=" + i);
                }
            }
        }
    }
```

and this will pass the test.

**Comment by plesingr-aeb** [comment]
Sure it will pass the test, but it does not solve the memory leak. We need to solve the leak.

At least, if we knew where the leak comes from (which package/class/method to look for it), we would definitely find a workaround. Could somebody, who knows the code a bit, help us, please?

**Comment by plesingr-aeb** [comment]
I have investigated the issue in more detail and found the problem in the `hk2-api` library - see https://github.com/eclipse-ee4j/glassfish-hk2/issues/949

**Comment by dmatej** [comment]
I think this issue can be closed, the fix was in HK2: https://github.com/eclipse-ee4j/glassfish-hk2/pull/965

**Comment by plesingr-aeb** [comment]
> I think this issue can be closed, the fix was in HK2: [eclipse-ee4j/glassfish-hk2#965](https://github.com/eclipse-ee4j/glassfish-hk2/pull/965)

Yes, but is the fix included in the jersey release (which one), too?

**Comment by senivam** [comment]
version of HK2 should be updated in the 3.1.6 version of Jersey, which will be released soon, however, if you use Jersey 3.1.x now, you can add ``-Dhk2.version=3.0.6`` to the build configuration to use fixed HK2 sooner.
